pkg/agent/proxy/proxier.go

```diff
@@ -485,49 +490,73 @@ func (p *proxier) installServices() {
 				bothPolicyLocal := internalPolicyLocal
 				groupID := p.groupCounter.AllocateIfNotExist(svcPortName, bothPolicyLocal)
 				if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, allReachableEndpoints); err != nil {
-					klog.ErrorS(err, "Error when installing Group of local Endpoints for Service", "Service", svcPortName)
+					klog.ErrorS(err, "Error when installing group of Endpoints for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr, "BothTrafficPolicies", bothPolicyLocal)
 					continue
 				}
 				if groupID, exist := p.groupCounter.Get(svcPortName, !bothPolicyLocal); exist {
 					if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
-						klog.ErrorS(err, "Failed to uninstall Group of all Endpoints for Service", "Service", svcPortName)
+						klog.ErrorS(err, "Error when uninstalling group of Endpoints for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr, "BothTrafficPolicies", !bothPolicyLocal)
 						continue
 					}
 					p.groupCounter.Recycle(svcPortName, !bothPolicyLocal)
 				}
 			}
 
-			for _, e := range allReachableEndpoints {
-				// If the Endpoint is newly installed, add a reference.
-				if _, ok := endpointsInstalled[e.String()]; !ok {
-					key := endpointKey(e, svcInfo.OFProtocol)
+			// Map endpointsInstalled may have stale Endpoints, we need to calculate a new map updatedEndpointsInstalled
+			// to replace it.
+			updatedEndpointsInstalled := map[string]k8sproxy.Endpoint{}
+			for _, endpoint := range allReachableEndpoints {
+				// Add the Endpoint to map updatedEndpointsInstalled since Endpoints in allReachableEndpoints are actually
+				// installed Endpoints.
+				updatedEndpointsInstalled[endpoint.String()] = endpoint
+
+				// If the Endpoint is not newly installed for the Service, remove it from map endpointsInstalled. If all
+				// the Endpoints that are not newly installed are removed from map endpointsInstalled, only stale Endpoints
+				// are left.
+				if _, exists := endpointsInstalled[endpoint.String()]; exists {
+					delete(endpointsInstalled, endpoint.String())
+				} else {
+					// If the Endpoint is newly installed, add a reference.
+					key := endpointKey(endpoint, svcInfo.OFProtocol)
 					p.endpointReferenceCounter[key] = p.endpointReferenceCounter[key] + 1
-					endpointsInstalled[e.String()] = e
 				}
 			}
+
+			// Update the Endpoints installed of the Service in endpointsInstalledMap.
+			delete(p.endpointsInstalledMap, svcPortName)
+			if len(updatedEndpointsInstalled) != 0 {
+				p.endpointsInstalledMap[svcPortName] = updatedEndpointsInstalled
+			}
```

Comment:
The same comment applies here. It shouldn't update cache before the datapath change succeeds

Reply:
Thanks for reminding, updated.